fix: add warmup for audio preprocessing

TheCodeWrangler opened the pull request:
# Audio Preprocessing Warmup for Whisper Models

## Purpose

Fixes first-request latency issue for Whisper transcription models where the first request can take significantly longer (10-20x) than subsequent requests.

**Problem**: The first Whisper transcription request experiences substantial latency overhead due to lazy initialization, causing poor user experience and potential timeout issues in production.

**Root Cause**: Investigation revealed two lazy initialization bottlenecks:
1. **Librosa/audio preprocessing**: First call to `librosa.load()` and mel-spectrogram functions triggers FFTW plan initialization, numba JIT compilation, and audio codec loading
2. **Input processor**: First call to `input_processor.process_inputs()` with multimodal audio triggers WhisperProcessor initialization and tensor operation JIT compilation

**Solution**: Add warmup methods that run during server initialization to eliminate first-request overhead:
- `_warmup_audio_preprocessing()`: Warms up librosa functions with dummy audio
- `_warmup_input_processor()`: Warms up multimodal processing pipeline with dummy request

## Test Plan

### Setup
```bash
# Start vLLM server with Whisper model
vllm serve openai/whisper-large-v3 --dtype bfloat16

# Run test client that makes 10 consecutive transcription requests
python testing/timed_requests_whisper.py
```

### Test Script
Sends 10 requests with the same 60-second audio file and measures client-side latency for each.

## Test Results

### Example Results

**Test Configuration:**
- Model: `openai/whisper-large-v3`
- Hardware: H100 GPU
- Audio: 60-second MP3 file
- Test: 10 consecutive transcription requests

#### Before Fix
```
Request 0: 15.56s  ← First request (18x slower!)
Request 1-9: ~0.85s each
```

**Server-side breakdown (first request)**:
- Audio preprocessing (librosa): **7.67s** (lazy initialization)
- Input processor (multimodal): **2.53s** (lazy initialization)
- Model execution: ~0.02s (already warmed up)
- Text generation: ~5.3s

#### After Fix
```
Request 0: 1.08s  ← First request (now similar to subsequent!)
Request 1-9: ~0.85s each
```

**Server-side breakdown (first request)**:
- Audio preprocessing: **0.08s** (warmed up during startup)
- Input processor: **0.007s** (warmed up during startup)
- Model execution: ~0.02s
- Text generation: ~0.5s

#### Impact Summary

| Component | Before | After | Improvement |
|-----------|--------|-------|-------------|
| Audio preprocessing | ~7-8s | ~0.04-0.08s | ~100-200x faster |
| Input processor | ~2-3s | ~0.002-0.007s | ~400-1500x faster |
| **First request** | ~15-20s | ~1-1.5s | **~10-20x faster** |
| Subsequent requests | ~0.8-1s | ~0.8-1s | No change |
| **Server startup** | Baseline | +10-12s | One-time cost |

**Note**: Exact timings vary based on model size, hardware, audio duration, and system configuration. The key improvement is eliminating the large first-request overhead caused by lazy initialization.

### Trade-off Analysis
- **Added startup time**: ~10-12s (warmup runs once during initialization)
- **Saved per first-request**: ~10-15s
- **Break-even**: After just 1 first request
- **User experience**: Consistent latency across all requests

## Implementation Details

### Changes Made

**File**: `vllm/entrypoints/openai/speech_to_text.py` (+95 lines)

1. **`_warmup_audio_preprocessing()`**
   - Creates 1-second dummy audio (silence at target sample rate)
   - Calls `librosa.get_duration()` to initialize librosa
   - Calls `librosa.feature.melspectrogram()` with Whisper-specific parameters
   - Initializes FFTW plans, numba JIT, audio codecs
   - Non-fatal: logs warning if warmup fails

2. **`_warmup_input_processor()`**
   - Creates dummy audio and generates proper prompt using `model_cls.get_generation_prompt()`
   - Creates minimal `SamplingParams`
   - Calls `input_processor.process_inputs()` to process dummy request
   - Triggers WhisperProcessor initialization, multimodal hashing, tensor operations
   - Non-fatal: logs warning if warmup fails

3. **Called in `__init__()`**
   - Both warmups run after audio config setup
   - Executed before server starts accepting requests
   - Logs clearly indicate warmup progress

### Why This Approach

- **Targeted**: Only warms up components that showed first-call overhead
- **Safe**: Error handling ensures server starts even if warmup fails
- **Observable**: Clear logging shows warmup completion and timing
- **Efficient**: Uses minimal resources (1-second dummy audio)
- **Complete**: Warms up entire pipeline end-to-end

### Alternate Approaches Considered

1. **Lazy warmup on first request**: Rejected - still causes bad UX for first user
2. **Background warmup**: Rejected - adds complexity, potential race conditions
3. **Only librosa warmup**: Insufficient - still left 2.5s input processor delay
4. **Disable JIT**: Rejected - would hurt performance on all requests

## Documentation Update

None required - this is an internal optimization with no API changes.

## Release Notes

**Performance Improvement**: Whisper transcription models now warm up audio preprocessing and multimodal input processing during server initialization, eliminating 10+ seconds of first-request latency. First requests are now 14x faster (15.5s → 1.08s), providing consistent latency for all users.

## Related Issues

This addresses first-request latency for audio/speech-to-text models, particularly Whisper. The issue manifests as:
- First transcription request timing out or taking 15+ seconds
- Poor user experience for cold-start scenarios
- Inconsistent latency patterns

## Testing Checklist

- [x] Tested with `openai/whisper-large-v3` model
- [x] Verified warmup completes successfully during server startup
- [x] Confirmed first request latency reduced from 15.5s to 1.08s
- [x] Verified subsequent requests unchanged at ~0.85s
- [x] Tested with 60-second audio files
- [x] Verified warmup logs appear during initialization
- [x] Confirmed non-fatal error handling (server starts even if warmup fails)
- [x] No functional changes to transcription accuracy or behavior
- [x] No changes to API surface or response format

## Additional Notes

- First request may still be slightly slower than subsequent requests (~10-30% overhead) due to first HTTP connection, first batch through scheduler, and first encoder hash computation - this is acceptable and unavoidable without significantly more complex warmup
- Warmup is automatic and requires no configuration changes
- Applies to all speech-to-text models using the OpenAI-compatible transcription API (Whisper, Voxtral, etc.)
- Server startup time increase is a one-time cost that provides immediate value on first user request
- Warmup uses minimal resources (1 second of silence) and completes in the background during initialization

TheCodeWrangler committed:
fix: Add audio preprocessing warmup to eliminate Whisper first-request latency

Adds warmup for librosa and input processor to eliminate 10+ seconds
of first-request latency for Whisper transcription models.

First request: 15.5s -> 1.08s (14.4x faster)
Subsequent requests: 0.85s (unchanged)
Server startup: +11.4s (one-time cost)

Signed-off-by: Nathan Price <nathan@abridge.com>

TheCodeWrangler committed:
Merge branch 'main' into fix-whisper-first-call-latency

TheCodeWrangler commented:
@LucasWilkinson @robertgshaw2-redhat

Related to prior warmup PR: https://github.com/vllm-project/vllm/pull/30561

But targeted to audio transcription models.

TheCodeWrangler committed:
fix: Add audio preprocessing warmup to eliminate Whisper first-request latency

Signed-off-by: Nathan Price <nathan@abridge.com>

TheCodeWrangler committed:
Fixed import bug

Signed-off-by: Nathan Price <nathan@abridge.com>

DarkLight1337 reviewed (commented):
We should avoid calling these methods unless the model actually supports it

TheCodeWrangler committed:
updated to check for feature extractor before using method

Signed-off-by: Nathan Price <nathan@abridge.com>

TheCodeWrangler committed:
Merge branch 'main' into fix-whisper-first-call-latency

TheCodeWrangler committed:
Merge branch 'main' into fix-whisper-first-call-latency

TheCodeWrangler committed:
Merge branch 'main' into fix-whisper-first-call-latency

TheCodeWrangler committed:
Merge branch 'main' into fix-whisper-first-call-latency

TheCodeWrangler committed:
Add defensive checks to skip warmup for non-multimodal models

- Skip audio preprocessing warmup if librosa is not installed (optional dependency)
- Skip warmup if model doesn't support multimodal/audio inputs
- Skip input processor warmup if model doesn't support get_generation_prompt
- Prevents unnecessary startup time and log pollution for models that don't use these features

Signed-off-by: Nathan Price <nathan@abridge.com>

TheCodeWrangler committed:
Merge branch 'main' into fix-whisper-first-call-latency

TheCodeWrangler committed:
Use supports_transcription instead of is_multimodal_model for warmup checks

- Replace is_multimodal_model check with supports_transcription check
- More specific and appropriate for speech-to-text context
- Only checks if model actually supports transcription rather than general multimodal support

Signed-off-by: Nathan Price <nathan@abridge.com>

DarkLight1337 reviewed (approved):
Thanks!

TheCodeWrangler committed:
Merge branch 'main' into fix-whisper-first-call-latency

TheCodeWrangler committed:
Merge branch 'main' into fix-whisper-first-call-latency

DarkLight1337 committed:
Merge branch 'main' into fix-whisper-first-call-latency

TheCodeWrangler committed:
Increased logger levels

Signed-off-by: Nathan Price <nathan@abridge.com>

DarkLight1337 commented on vllm/entrypoints/openai/speech_to_text.py:
```diff
@@ -233,7 +233,7 @@ def _warmup_input_processor(self) -> None:
         except Exception as e:
             # Don't fail initialization if warmup fails - log warning and continue
-            logger.warning(
+            logger.exception(
```
In that case there is no need to log `e` since `exc_info=True` automatically

TheCodeWrangler committed:
Removed explict addition of exception in logger.exception

Signed-off-by: Nathan Price <nathan@abridge.com>

TheCodeWrangler committed:
Merge branch 'main' into fix-whisper-first-call-latency

TheCodeWrangler committed:
Merge branch 'main' into fix-whisper-first-call-latency

TheCodeWrangler committed:
Merge branch 'main' into fix-whisper-first-call-latency

DarkLight1337 committed:
Merge branch 'main' into fix-whisper-first-call-latency
